### PR TITLE
Add a Shutdown() method to `*Provider` SDK

### LINKED ISSUE #1073: Add a Shutdown() method to `*Provider`
**What are you trying to achieve?**

Add a Shutdown() method to `TracerProvider` and `MeterProvider`, which would invoke Shutdown() on all processors.

**Additional context.**

There is no convenient way to let our users shut the processors down if their programs want to exit. It is the users' response to keep `Shutdown` functions from processors. Telemetry API should handle this common scenario, which every program does, to make sure spans/metrics are written before exit.

Related https://github.com/open-telemetry/opentelemetry-go/issues/1206

### PR DESCRIPTION
**Pull request by XSAM**
Resolve #1073

### CODE REVIEW

File: specification/metrics/api.md [hunk 1]
```diff
@@ -358,6 +358,12 @@ other libraries.  The name of the `Meter` is explicitly not intended
 to be used as part of the instrument name, as that would prevent
 instrumentation libraries from capturing metrics by the same name.
 
+### Shutdown
+
+Exporting pending metrics.
+
+The `Shutdown` API MAY be implemented by invoking `Shutdown` within internal processors.
```
Comment (tigrannajaryan) [inline specification/metrics/api.md:365]: This is an implementation details. I am not sure we want this in the API spec. I think it is better to delete this sentence or move to the sdk.md.

File: specification/metrics/api.md [hunk 2]
```diff
@@ -358,6 +358,12 @@ other libraries.  The name of the `Meter` is explicitly not intended
 to be used as part of the instrument name, as that would prevent
 instrumentation libraries from capturing metrics by the same name.
 
+### Shutdown
+
+Exporting pending metrics.
```
Comment (tigrannajaryan) [inline specification/metrics/api.md:363]: Is it possible to specify this more precisely? What do we consider a "pending" metric? The term "pending" is not described anywhere in the metric API spec. Does this initiate a collection of asynchronous instruments?

File: specification/trace/api.md [hunk 3]
```diff
@@ -125,6 +126,8 @@ the tracer could, for example, do a look-up with its name+version in a map in
 the `TracerProvider`, or the `TracerProvider` could maintain a registry of all
 returned `Tracer`s and actively update their configuration if it changes.
 
+The `Shutdown` API MAY be implemented by invoking `Shutdown` within internal processors.
```
Comment (tigrannajaryan) [inline specification/trace/api.md:129]: Similarly here, I think we don't need this in the API spec.

File: specification/trace/api.md [hunk 4]
```diff
@@ -91,11 +91,12 @@ number of `TracerProvider` instances.
 
 ### TracerProvider operations
 
-The `TracerProvider` MUST provide functions to:
+The `TracerProvider` MUST provide the following functions:
 
 - Get a `Tracer`
+- Shutdown the provider (exporting pending spans)
```
Comment (tigrannajaryan) [inline specification/trace/api.md:97]: Similarly here: what is a "pending" span?

File: specification/trace/api.md [hunk 5]
```diff
@@ -91,11 +91,12 @@ number of `TracerProvider` instances.
...
 ### TracerProvider operations
 
-The `TracerProvider` MUST provide functions to:
+The `TracerProvider` MUST provide the following functions:
 
 - Get a `Tracer`
+- Shuts down the provider
 
-That API MUST accept the following parameters:
+That `Tracer()` API MUST accept the following parameters:
```
Comment (Oberon00) [inline specification/trace/api.md:99]: There is never such a (named) API introduced. Better restructure this a bit. E.g. write above "The TracerProvider MUST provide the functionality described in the following subsections" and then make a subheading each for `#### Get a Tracer` and `#### Shutdown`

File: specification/trace/sdk.md [hunk 6]
```diff
@@ -180,7 +180,9 @@ Optional parameters:
 |present|false|true|`localParentSampled()`|
 |present|false|false|`localParentNotSampled()`|
 
-## Tracer Creation
+## Tracer Provider
```
Comment (Oberon00) [inline specification/trace/sdk.md:183]: 👍

File: specification/trace/sdk.md [hunk 7]
```diff
@@ -201,6 +203,22 @@ Note: Implementation-wise, this could mean that `Tracer` instances have a
 reference to their `TracerProvider` and access configuration only via this
 reference.
 
+### Shutdown
+
+`Shutdown` SHOULD be called only once for each `TracerProvider` instance. After
+the call to `Shutdown`, subsequent calls to get `Tracer` are not allowed. SDKs
```
Comment (Oberon00) [inline specification/trace/sdk.md:209]:
```suggestion
the call to `Shutdown`, subsequent attempts to get a `Tracer` are not allowed. SDKs
```

File: specification/trace/sdk.md [hunk 8]
```diff
@@ -201,6 +203,22 @@ Note: Implementation-wise, this could mean that `Tracer` instances have a
...
+SHOULD ignore these calls gracefully, if possible.
```
Comment (Oberon00) [inline specification/trace/sdk.md:210]:
```suggestion
SHOULD return a valid no-op Tracer for these calls, if possible.
```

We can be a bit more specific here.

File: specification/trace/sdk.md (hunk 7 above)
Comment (Oberon00) [inline specification/trace/sdk.md:209]: I wonder if placing this requirement on the user is OK. There may be asynchronous threads running that still produce spans / get tracers while the telemetry SDK shuts down. I think we should make it a MUST to ignore subsequent attempts to get a tracer.

I think it's OK to only allow one call to shutdown though.

File: specification/trace/sdk.md [hunk 9]
```diff
@@ -201,6 +203,22 @@ Note: Implementation-wise, this could mean that `Tracer` instances have a
...
+
+`Shutdown` SHOULD provide a way to let the caller know whether it succeeded,
+failed or timed out.
+
+`Shutdown` SHOULD complete or abort within some timeout. `Shutdown` can be
+implemented as a blocking API or an asynchronous API which notifies the caller
+via a callback or an event. Language library authors can decide if they want to
+make the shutdown timeout configurable.
+
+The `Shutdown` API MAY be implemented by invoking `Shutdown` within internal processors.
```
Comment (Oberon00) [inline specification/trace/sdk.md:220]: I think we need to be more strict here. Shutdown MAY do other things but at the very least it MUST invoke shutdown on all contained processors.

**Review by Oberon00 — COMMENTED** [review]
Thank you for the PR! I'm definitely in favor of adding Shutdown

**Comment by arminru** [comment]
@XSAM Please add the respective entries in CHANGELOG.md and spec-compliance-matrix.md. Thanks!

File: specification/trace/api.md [hunk 10]
```diff
@@ -91,11 +91,14 @@ number of `TracerProvider` instances.
 
 ### TracerProvider operations
 
-The `TracerProvider` MUST provide functions to:
+The `TracerProvider` MUST provide the following functions:
 
 - Get a `Tracer`
+- Shuts down the provider
```
Comment (Oberon00) [inline specification/trace/api.md:97]: On second thought, I agree with @bogdandrutu's https://github.com/open-telemetry/opentelemetry-specification/issues/1073#issuecomment-706259426: We should not have this in the API, only the SDK.

**Review by Oberon00 — APPROVED** [review]
Thank you!

**Review by carlosalberto — APPROVED** [review]

**Review by arminru — APPROVED** [review]

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**